### ISSUE TITLE
Sigma Static Bridge Controller support (eg Silabs UZB-7)

### ISSUE THREAD

**Issue by m-p-goodwin** [OpenZWave/open-zwave#1768]
I guess I could comment on some of #1072, #1345, #615, #739 or #953, but these have all been closed long ago, and I guess necroposting is also frowned upon here.

I would also like to use the UZB stick for my zwave network. I can see from #739 that it would require reverse engineering of the protocol, but I found (google) a document of the protocol, released in 2018, which is why it was not available at the time of the other issues. I attach it here (if I can - it is 1.3MB)
[INS12350-Serial-API-Host-Appl.-Prg.-Guide.pdf](https://github.com/OpenZWave/open-zwave/files/3088295/INS12350-Serial-API-Host-Appl.-Prg.-Guide.pdf)

**Comment by Fishwaldo** [comment]
I'm aware of that document, and that most of the Specs are now released (or at least available, its not clear from SiLabs on the status of them). But That SerialAPI document (and the other documentation available is very lightweight on Bridge vs Controller etc. (eg, the SerialAPI document doesn't detail what the message type 0xA8 is!

But some source files are available, so it should be possible for someone to run through them and figure it out. 

Good news for you though is that there are alternative firmware versions available for the UZB, so you can switch from Bridge to Static Controller and be able to use it with OZW.

**Comment by m-p-goodwin** [comment]
Interesting - How is this actually done, without necessarily getting all the tools and images off of Silicon Labs page, under NDA.

**Comment by petergebruers** [comment]
If you register on the Silabs website you can download the docs and tools.

FUNC_ID_APPLICATION_COMMAND_HANDLER_BRIDGE = 0xA8

ZW->HOST: REQ | 0xA8 | rxStatus | destNodeID | srcNodeID | cmdLength | pCmd[ ] | multiDestsOffset_NodeMaskLen | multiDestsNodeMask | rxRSSIVal

"destNodeID == Command receiving Node ID. Either Bridge Controller Node ID, virtual slave Node ID or if in promiscuous mode Node ID on node belonging to same HomeID as current node.

See 4.3.1.9 ApplicationCommandHandler_Bridge

in INS13954 "Z-Wave 500 Series Appl. Programmers Guide v6.81.0x"

Preliminary testing tels me it basically replaces the 0x04 because a gateway needs to be able to report to any node ID, not only the controller.

I have a UZB1 running gateway software, I can test it.

**Comment by m-p-goodwin** [comment]
I sort of gathered that I can get the necessary tools through their homepage, but also thought I should check if there is a different way to flash the uzb stick - A way that sidesteps the NDA that you are asked to approve when asking to download the programmer. (That you have obviously agreed to, considering your "fair use" quote. No problem, I am also considering, but would like to hear if there are other options.)

**Comment by petergebruers** [comment]
In case of the Z-Wave.Me UZB1 or RaZberry you install their Z-Way gateway and upgrade the stick, they have released both controller and bridge firmware for their device.

https://z-wave.me/support/uzbrazberry-firmwares/

In my case I don't need the programmer and I don't violate the NDA.

I don't know what to advice, in case you have another brand... Maybe buy a Silabs UZB-7? But in that case, you might have a similar problem... You would have to download a firmware image and flash it (it might come empty, I don't own one and cannot check) - which you would obtain through their website.

**Comment by Fishwaldo** [comment]
Ok... For Clarity - Are we talking about the zway.me stick or the Sigma/Silabs Stick?

For the Sigma/Silabs stick - instructions here on how to update the firmware: https://www.silabs.com/community/wireless/z-wave/knowledge-base.entry.html/2018/12/19/z-wave_500_programm-BQRO

Regarding actually implementing Bridge Firmware Support - It would be low on my priorities as there are Static Controller Firmware available, so their is a easier workaround. :)

**Comment by petergebruers** [comment]
I own the UZB1 made by Z-Wave.me and have one running "controller" lib and another one "bridge" lib.
It might not be urgent, but AFAIK the new Silabs UZB called UZB-7 (so based on 700) only has "gateway" lib. I know, from a friend, OpenHAB is incompatible with this new stick and in the log file he sees data coming in as FUNC_ID_APPLICATION_COMMAND_HANDLER_BRIDGE. He does not run OZW and I don't have a UZB-7...

I also own a "Aeotec Z-Stick ZW5" and I flashed it with Zniffer firmware so I suppose I could flash it with bridge firmware as well (not sure if it would work 100% because there is an amplifier on the board... But is seems to work as sniffer). You have to be careful though, it is possible to brick the device, eg by flashing a firmware that does not support USB. In that case you have to solder tiny wires to the board. I have done that, but it is not easy... AFAIK I cannot flash the Aeotec stick with stock firmware, unless I can get the original firmware from them. I never tried, because I wanted this to be a Zniffer anyway.

**Comment by petergebruers** [comment]
I've hacked in bridge support based on OZW 1.6 (master). It works on UZB-7 and on a "Aeotec Z-Stick ZW5" that has the reference implementation of the SK 6.81 bridge firmware + an antenna modification.  The test network is small, only 2 relays. Basic functionality with ozwcp and hass is there, this is enough for me to show that it is possible to support bridge library. Update neighbor seems broken, I haven't had time to investigate. If you want to try it, use the "bridge" branch of my fork:

https://github.com/petergebruers/open-zwave

**Comment by petergebruers** [comment]
Merged 1.6 master branch into my bridge branch. Master has several fixes and improvements.
Tested with open-zwave-control-panel and 1x FGS221 double relay.

**Comment by BilliAlpha** [comment]
Hi @petergebruers, Is you bridge branch stable ?
Could it be used with the zwave2mqtt tool ?

**Comment by petergebruers** [comment]
The bridge branch is based on the master branch of OpenZWave (though I have not merged upstream in a while) so it is as stable as the OpenZWave 1.6 version. At the moment, zwave2mqtt is not fully compatible with OpenZWave 1.6, there is at least this issue with bitsets

https://github.com/OpenZWave/Zwave2Mqtt/issues/94

I cannot tell much about the Zwave2Mqtt project, I do not use it myself.

@BilliAlpha I am curious, what kind of controller would you want to use with the bridge branch?

**Comment by BilliAlpha** [comment]
Hi, thanks for taking the time to reply.

I'm using the sigma designs bridge controller (ACC-UZB3-E-BRG).

I often see errors in the logs regarding the 0xA8 packet so I hopped it would work better with bridge support.

**Comment by petergebruers** [comment]
> I'm using the sigma designs bridge controller (ACC-UZB3-E-BRG).

OK.

> I often see errors in the logs regarding the 0xA8 packet so I hopped it would work better with bridge support.

Yes! That is what "bridge support" means, it is all about that 0xA8 packet. 🙂 

If you checkout my "bridge" branch you will see in Driver.cpp:

```
void Driver::HandleApplicationCommandHandlerBridgeRequest
(
		uint8* _data,
		bool encrypted
)
{
	// Frame: request, 0xA8, rxStatus, destNode, sourceNode, cmdLength, pCmd, multiDestsOffset_NodeMaskLen,
	// multiDestsNodeMask, rxRSSIVal

	// Non-bridge version: request, 0x04, rxStatus, sourceNode, cmdLength, pCmd[], rxRSSIVal, securityKey
```

There is your 0xA8 aka FUNC_ID_APPLICATION_COMMAND_HANDLER_BRIDGE

So that branch handles both normal controller firmware and bridge firmware.

**Comment by BilliAlpha** [comment]
Ok, thanks.
I'll try it out a bit but I probably should wait for the bitset issue to be solved.

**Comment by Hedda** [comment]
FYI, here are a couple more links about Silicon Labs UZB-7 (Z-Wave 700 series controller USB dongle):

https://www.silabs.com/documents/public/data-sheets/DSH14483.pdf

https://fccid.io/2ADB9-UZB-7/User-Manual/Users-Manual-4147020

Again, this is Silicon Labs reference design of a Z-Wave 700 series controller USB dongle based on Silicon Labs EFR32ZG14 "Zen Gecko" Z-Wave 700 modem System-on-Chip.

**Comment by Kahuna711** [comment]
Hi,

I was able to buy a " Z-Wave 700 UZB-7 USB Stick" (SLUSB001A) at Mouser for just 17,55€. Delivery time 3 days...

**Comment by petergebruers** [comment]
Thanks @Kahuna711 

Advice to other readers. Before you buy one, please make sure you understand this:

- Support in OZW is experimental and may have bugs. We need testers and possibly coders to get it to an acceptable level. It is not sure if it will ever make it into "master". That depends on code quality and feedback.
- The UZB-7 only comes with "gateway" firmware. There is no "controller" firmware (the "old" type of firmware) and Silabs has stated it will never make such a firmware. So it won't be compatible with existing software (unless that software has support for gateway firmware which IMHO only applies to Z/IP). There also is no "Zniffer" firmware and it is unclear if there will ever be. If you buy the Silabs SDK, you get a series 500 dongle to use as a Zniffer.
- All official communication says the series 700 controller functionality is exposed by Z/IP only 'Z/IP compiles on many platforms, I use it for testin on my Mac). There is no "serial protocol definition" for the series 700. That means we actually base ourselves on what is known about the 500 series serial protocol and reverse engineering.
- Given the statement about Z/IP, it is unclear if there will ever be a "dongle" with documented serial protocol, either manufactured by Silabs or a 3rd party. Technically it would be possible, but someone suggested it would not be possible to certify such a device. If you really  want to find out all the details, be my guest to read the docs or contact Silabs. We are a bit understaffed here...

**Comment by Fishwaldo** [comment]
This is closed via #2126
